Test framework update

JonathanBurke opened the pull request:
The test framework is replaced.  The new class to extend is DefaultCheckerTest.  
If you have a custom test, you can always use the test utilities to easily run the framework with a given set of options.

JonathanBurke commented on checker/tests/src/tests/IGJTest.java:
```diff
@@ -1,33 +1,45 @@
  * JUnit tests for the IGJ annotation checker.
  */
-public class IGJTest extends CheckerTest {
+public class IGJTest {
```
The IGJ Checker was the only test that didn't fit the usual mold.  It required a few of the old methods from the test framework which have been moved here.  It should be updated to match our testing conventions.

wmdietl replied:
I agree that the IGJ tests should be adapted to the current style - in particular merging all the .out files into the test .java files.

wmdietl commented on checker/build.xml:
```diff
@@ -510,6 +520,15 @@
     <!-- Used to guarantee that targets executed from other ant scripts do not cause unnecessary rebuilds -->
     <target name="all-tests-nojar" depends="build-tests"
             description="Run tests for all checkers, WITHOUT building anything">
+
```
Can we reduce the duplication, which existed before your changes, and somehow merge the two "junit" tasks?

JonathanBurke replied:
I agree with this.  I just haven't had time since I am trying to focus on the Verigames experiment we're trying to run.  I'll see if this can be done quickly but if not, I will add an issue and move on.

wmdietl replied:
How about adding a comment to the two duplicates to make sure anyone who edits one also edits the other?
And adding an issue that we should fix the duplication.

wmdietl commented on checker/tests/src/tests/FenumTest.java:
```diff
@@ -16,7 +17,7 @@ public FenumTest(File testFile) {
     @Parameters
-    public static Collection<Object[]> data() {
-        return testFiles("fenum", "all-systems");
+    public static Collection<Object[]> getTestFiles() {
```
This method repeats a lot, in each test file, and only the directory names differ.
To reduce this duplication and possible source of mistakes, could we introduce a

String[] getTestDirectories()

method? Then one would only need to write

return {"fenum", "all-systems"};

(If array literals work in that context...)

JonathanBurke replied:
That's a good suggestion.  I have done the following:

DefaultCheckerTest nows asks you to specify one of two methods:
```
@Parameters
public static void List<Object[]> getTestFiles() 
```
Which does the old behavior and allows people to find arbitrary test files (which is used in the InferenceTest suite to find all-systems in the checker-framework directory).
OR users can specify
```
@Parameters
public static void String[] getTestDirs() 
```
Which will return a list of directories that should be in the test directory.  The TestSuite will now enforce that a single method with one of the exact signatures above and an `@Parameters` has been added to your test file and otherwise it will explode with an informative exception/message.

JonathanBurke replied:
Note, I think in the long run we should separated the tests and their corresponding input test files into two directories:
test 
testdata

I think it's confusing to have both the testdata and the tests that run against that data in the same folder at the same level.

wmdietl commented on framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java:
```diff
@@ -0,0 +1,81 @@
+ * 2) Create a public static method that is annotation with org.junit.runners.Parameterized.Parameters.  This method
+ * should look like:
+ * {@code, @Parameters List<Object[]> getTestFiles(); }
+ * The return type should be a List of 1 element arrays, where each member of that array is a Java file to be tested.
```
Why is it a "List of 1 element arrays", but then "each member of that array"?

JonathanBurke replied:
Well I guess it should be:
I can update this comment. 
Alternatively:  I kept this signature because it matches the original framework.  We could make this a List<String> and do the conversion to a List<Object []> in the TestSuite to make this less ambiguous.    It needs to be List<Object []> because that's what JUnit expects for parameters.

wmdietl commented on framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java:
```diff
@@ -0,0 +1,81 @@
+ * the run method.
+ */
+@RunWith(TestSuite.class)
+public abstract class DefaultCheckerTest {
```
Should this be CheckerFrameworkTest?
Why the "Default" if it doesn't implement a "CheckerTest" interface?

JonathanBurke replied:
Well default in the sense that this is the default class users will want to extend.  Whereas, if you had special behavior you could just not extend this class and create your own `@Test` methods.  For instance, we also have an InferenceCheckerTest.

wmdietl commented on framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java:
```diff
@@ -0,0 +1,81 @@
+     * Creates a new checker test.
+     *
+     * @param checker the class for the checker to use
+     * @param checkerDir the path to the directory of test inputs
```
Why is the "checkerDir" parameter needed? Don't we get the test inputs from the "getTestFiles" method?

wmdietl replied:
Can we remove the parameter?

JonathanBurke replied:
So checkerDir is used right now to put on the sourcepath for the JUnit test.  We probably can but I think I will avoid doing this for now as we're getting close to release and I have many tasks.  Perhaps when open the issue for separating the Testdata I will also add information about removing this parameter.

wmdietl commented on framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java:
```diff
@@ -0,0 +1,87 @@
+public class ImmutableTestConfiguration implements TestConfiguration {
+
+    /**
+     * Options that should be passed to the compiler
```
Why is it a Map? What are the keys and values?

JonathanBurke replied:
Some options have arguments and some do not, e.g.,
-AprintAllQualifiers
vs.
-Aclasspath "dir1:dir2"
While you can argue that second case is more like 2 arguments, the point of the map is so I can do things like:

addToClasspath("dir1")
addToClasspath("dir2")

rather than
setClasspath("dir1")
setClasspath(getClasspath() + File.pathSeparator" + "dir2")

It is a pretty naive approach to adding options but works pretty well in most cases.  Since we are using it in a limited set of cases in the test suite I am not sure we need to make it any more sophisticated.

JonathanBurke replied:
Though I should probably change this to SimpleOptionMap

JonathanBurke replied:
I take that back.  The reason this is a map is because of the above reasoning but the ImmutableTestConfiguration should not contain the SimpleOptionMap (because it is mutable).

wmdietl replied:
You should add documentation for what the key and the values are.
I didn't make the connection from arguments to parameters being the values that the parameters map to.

JonathanBurke replied:
SimpleOptionMap is a utility class that contains a Map like above and a set of utility methods that manipulate the map.  SimpleOptionMap contains methods similar to addToClasspath above except they work like:  addToPath("-cp", "myClassesDir").  Then the ConfigurationBuilder's method looks as follows:
public void addToClasspath(String classesDir) {
   optionMap.addToPath("-cp", "myClassesDir");
}

JonathanBurke replied:
One final note, I have been meaning to add this documentation to the interface NOT the implementation, so most of these methods will just `@inherit` the Javadoc.

wmdietl commented on framework/src/org/checkerframework/framework/test/SimpleOptionMap.java:
```diff
@@ -0,0 +1,90 @@
+ * Created by jburke on 7/7/15.
+ */
+public class SimpleOptionMap {
+    private Map<String, String> options = new LinkedHashMap<>();
```
Field can be final?

wmdietl commented on framework/src/org/checkerframework/framework/test/SimpleOptionMap.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.Map;
+
+/**
+ * Created by jburke on 7/7/15.
```
Some documentation would be useful.

JonathanBurke replied:
Yeah, sorry.  I was just butting up against inference deadlines but I think I can add that now as that was pushed back a bit (but is still a top priority).

wmdietl commented on framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    public <T extends Collection<E>, E>  T addAll(T coll, Iterable<? extends E> objs) {
+        for(E obj : objs) {
```
Spacing in this file should consistently use "for (".

wmdietl commented on framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java:
```diff
@@ -0,0 +1,363 @@
+        shouldEmitDebugInfo = false;
+    }
+
+    public <T extends Collection<E>, E>  T addAll(T coll, Iterable<? extends E> objs) {
```
Should this method be static? private?
The method modifies the first parameter. None of the invocations of the method that I see in this diff use the returned value. It would be clearer to return void and document the modification of the parameter.

JonathanBurke replied:
This method should actually be removed.  There is one below that is private static that is used instead.

JonathanBurke replied:
That method, catIterableAndList has had its name changed to catListAndIterable

wmdietl commented on framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    /**
+     * This is creates a builder for the default configuration used by Checker Framework JUnit tests.
```
Remove "is".

wmdietl commented on framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java:
```diff
@@ -0,0 +1,363 @@
+
+    private boolean shouldEmitDebugInfo;
+
+    public TestConfigurationBuilder() {
```
Should this constructor be private to force usage of the static factory methods?
Should those factory methods be moved to the top of the file to make them easier to find?

JonathanBurke replied:
I did not plan on forcing people to use the static methods.  It can be used either way.  I just added those methods in order to make it easy to cover the most common case.  I can move them up top, however.

JonathanBurke replied:
Thinking this over, I think moving those static methods to the top makes the

wmdietl commented on framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java:
```diff
@@ -0,0 +1,363 @@
+        return coll;
+    }
+
+    public TestConfigurationBuilder(TestConfiguration initialConfig) {
```
This constructor doesn't seem to be used. Should it be removed? private?

JonathanBurke replied:
This constructor is used in the Inference Framework.

wmdietl commented on checker/tests/src/tests/OIGJTest.java:
```diff
@@ -21,8 +22,8 @@ public OIGJTest(File testFile) {
+        return filter(TestUtilities.findNestedJavaTestFiles("oigj", "all-systems"));
     }
 
     // Duplicate from JavariTest.
```
Is it worth moving the filter functionality into TestUtilities?
No, actually we should remove that code here and skip-test the files that are filtered.

JonathanBurke replied:
I agree.  Perhaps we can put an issue to fix both the IGJ/OIGJ tests together and somebody can handle that when they get a chance.  However, I worry that it's low priority so it might never get done.  Then again, at least this would be an easy task for a new hire, one in which they could make forward progress right away.

wmdietl commented:
Thanks for tackling this beast! I reviewed about the first half of the changes. I will continue with the rest some other time, maybe after you updated it.

JonathanBurke commented:
Hi Werner, I have incorporated most of your feedback (and my subsequent replies), added more documentation, and removed a couple of superfluous methods.  Please let me know if you have any other suggestions.

wmdietl commented on checker/build.xml:
```diff
@@ -521,13 +540,15 @@
       <mkdir dir="${build.reports}"/>
 
-      <junit fork="${run.tests.should.fork}"
+        <junit fork="${run.tests.should.fork}"
```
Unnecessary whitespace change.

JonathanBurke replied:
the spacing on this target is actually off compared to the others in the doc.  I will make it consistent.

wmdietl commented on framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java:
```diff
@@ -0,0 +1,88 @@
+ *
+ */
+@RunWith(TestSuite.class)
+public abstract class DefaultCheckerTest {
```
I would still call it CheckerFrameworkTest, because it's not a default implementation for some interface.
But this is OK. (I saw a comment from you on this in email, but not here...)

wmdietl commented on framework/src/org/checkerframework/framework/test/TestConfiguration.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Map;
+
+/**
+ * Created by jburke on 6/24/15.
```
A one sentence summary of the purpose of the interface would be more useful.

wmdietl commented on framework/src/org/checkerframework/framework/test/TestSuite.java:
```diff
@@ -0,0 +1,200 @@
+        return method;
+    }
+
+    private static String requiredFormsMessage =
```
Can be final?

wmdietl commented:
Great job! Having getTestDirs makes tests so much simpler!
I think we should pull this and get it into the upcoming release.
